### PR TITLE
Clear string-tracking hash table when ByteView deduplication is enabled

### LINKED ISSUE #6384: StringViewBuilder with deduplication does not clear observed values
**Describe the bug**
As of Arrow 53.0.0 as well as on mainline commit f050ff7a1946858d500b21a2c7d1b2c24fc2c753, I can cause a panic by reusing a `StringViewBuilder` that has deduplication enabled if I append a previously-observed value after having called `StringViewBuilder::finish` instead of `finish_cloned`. 

**To Reproduce**

```rust
let mut builder = StringViewBuilder::new().with_deduplicate_strings();
let value_1 = "long string to test string view";
builder.append_value(value_1);
let _array = builder.finish();
builder.append_value(value_1);
let _array = builder.finish();
```

**Expected behavior**

`StringViewBuilder::finish()` should leave me with a clean state with which I may continue to append new values, even if they have been previously observed **by this builder**.

**Additional context**

<details><summary>Possible patch, pending maintainer feedback in the PR I will open with the same</summary>

```diff
diff --git a/arrow-array/src/builder/generic_bytes_view_builder.rs b/arrow-array/src/builder/generic_bytes_view_builder.rs
index deaf447d..3a9cf17c 100644
--- a/arrow-array/src/builder/generic_bytes_view_builder.rs
+++ b/arrow-array/src/builder/generic_bytes_view_builder.rs
@@ -368,6 +368,9 @@ impl<T: ByteViewType + ?Sized> GenericByteViewBuilder<T> {
         let len = self.views_builder.len();
         let views = ScalarBuffer::new(self.views_builder.finish(), 0, len);
         let nulls = self.null_buffer_builder.finish();
+        if let Some((ref mut ht, _)) = self.string_tracker.as_mut() {
+            ht.clear();
+        }
         // SAFETY: valid by construction
         unsafe { GenericByteViewArray::new_unchecked(views, completed, nulls) }
     }
@@ -590,6 +593,20 @@ mod tests {
         assert_eq!(array.views().get(1), array.views().get(5));
     }
 
+    #[test]
+    fn test_string_view_deduplicate_after_finish() {
+        let mut builder = StringViewBuilder::new().with_deduplicate_strings();
+
+        let value_1 = "long string to test string view";
+        let value_2 = "not so similar string but long";
+        builder.append_value(value_1);
+        let _array = builder.finish();
+        builder.append_value(value_2);
+        let _array = builder.finish();
+        builder.append_value(value_1);
+        let _array = builder.finish();
+    }
+
     #[test]
     fn test_string_view() {
         let b1 = Buffer::from(b"world\xFFbananas\xF0\x9F\x98\x81");
```
</details>

### PR DESCRIPTION
**Pull request by shanesveller**
# Which issue does this PR close?

Closes #6384.

# Rationale for this change
 
The panic observed in #6384 was within `StringViewBuilder::get_value`, because the `string_tracker` hash table was not reset as part of `finish()`:

https://github.com/apache/arrow-rs/blob/e838e62d197cf0f6f653aa47aa00b5e5fb0b5f11/arrow-array/src/builder/generic_bytes_view_builder.rs#L365-L373

If the builder saw continued use, and `append_value` was called with a previously-observed value, the hash table would return an index to a now-emptied buffer after slicing into it here:

https://github.com/apache/arrow-rs/blob/e838e62d197cf0f6f653aa47aa00b5e5fb0b5f11/arrow-array/src/builder/generic_bytes_view_builder.rs#L257-L259

# What changes are included in this PR?

Builders will now clear the `string_tracker` hash-table as part of a revised `finish` definition.

# Are there any user-facing changes?

Not to my knowledge.

### CODE REVIEW

**Comment by shanesveller** [comment]
ℹ️  This is my first PR to the Arrow repository as well as to any Apache repository. I've reviewed the CONTRIBUTING documents as well as I can after watching day 1 of RustConf, but if I've overlooked anything or need to make any adjustments to the implementation, commit message, etc. to align with the project standards, I'm very happy to do so.

File: arrow-array/src/builder/generic_bytes_view_builder.rs [hunk 1]
```diff
@@ -590,6 +593,20 @@ mod tests {
         assert_eq!(array.views().get(1), array.views().get(5));
     }
 
+    #[test]
```
Comment (shanesveller) [inline arrow-array/src/builder/generic_bytes_view_builder.rs:596]: Please let me know if a regression test isn't needed for a simple change like this, I'm fine with dropping it if that's preferred.

**Review by tustvold — APPROVED** [review]
All looks good to me, thank you

**Comment by alamb** [comment]
FYI @XiangpengHao

**Comment by XiangpengHao** [comment]
look good to me, nice catch!

File: arrow-array/src/builder/generic_bytes_view_builder.rs (hunk 1 above)
Comment (alamb) [reply arrow-array/src/builder/generic_bytes_view_builder.rs:596]: Regression tests are *always* appreciated and we try to make then required.

Thank you for doing this @shanesveller

**Review by alamb — COMMENTED** [review]
❤️  Thanks @shanesveller @tustvold and @XiangpengHao

**Comment by alamb** [comment]
> ℹ️ This is my first PR to the Arrow repository as well as to any Apache repository. 

Welcome and thank you

**Pull request merged**